### PR TITLE
multistage builds draft language

### PR DESCRIPTION
**Pull request by ethangj**
@rheinwein @mlocher 

first draft, for review and editing

obviously needs the filler on caching filled out, as well

### CODE REVIEW

File: _pro/builds-and-configuration/caching.md [hunk 1]
```diff
@@ -87,3 +86,9 @@ The various RUN commands should also be ordered according to frequency of invali
 ### 3. Use a strict .dockerignore file
 
 The more files which get added to the Docker image during an ADD or COPY, the higher the chance that the image cached will be invalidated despite the functionality of the image remaining the same. To reduce the chances of this happening, strip down the number of files being added to the image to the bare essentials. Ignore any temporary files, logs, development files, and documentation, especially `.git`. A good rule of thumb for this process is if the resulting image will not utilize a file or folder, add it to the `.dockerignore` file.
+
+## Caching With Multi-stage Builds
+
+Docker's multi-stage build feature allows you to build Docker images with multiple build stages in the Dockerfile, ultimately saving an image from just the final stage. You can [read more about Docker multi-stage builds on our blog](link to laura's blog), but this has certain impacts on caching your image with Codeship Pro.
+
+You can [read more about Docker multi-stage builds on our blog](link to laura's blog).
```
Comment (mlocher) [inline _pro/builds-and-configuration/caching.md]: This is already duplicated in the paragraph above.

File: _pro/languages-frameworks/go.md [hunk 2]
```diff
@@ -77,6 +77,34 @@ COPY . ./
 RUN go get
 ```
 
+### Multi-stage Builds
+
+Using Docker's multi-stage build feature, you can implement some changes to you Dockerfile to allow you to build and use a Go binary from a single Dockerfile, outputting a Docker image with the Go binary but none of the Golang build tools - meaning a smaller and more efficient image with a less complex setup.
+
+Multi-stage builds allow you to specify multiple `FROM` lines in a Dockerfile, where each `FROM` line begins a new stage. The image saves with _the result of the last stage_, which means any previous stages are not saved in the final image. This is great for creating "builder" workflows easily.
```
Comment (mlocher) [inline _pro/languages-frameworks/go.md]: I would rewrite the second sentence as

> The final image is _the result of the last stage_, ...

File: _pro/languages-frameworks/java.md [hunk 3]
```diff
@@ -116,6 +116,30 @@ RUN gradle assemble
...
+
+Using Docker's multi-stage build feature, you can implement some changes to your Dockerfile to allow you to build and use a Java binary from a single Dockerfile, outputting a Docker image with the JAVA binary but none of the build tools - meaning a smaller and more efficient image with a less complex setup.
+
+Multi-stage builds allow you to specify multiple `FROM` lines in a Dockerfile, where each `FROM` line begins a new stage. The image saves with _the result of the last stage_, which means any previous stages are not saved in the final image. This is great for creating "builder" workflows easily.
+
+Here's an example using Java in a Dockerfile:
+
+```
+# phase one, labeled as BUILD
+FROM maven:latest as BUILD
```
Comment (mlocher) [inline _pro/languages-frameworks/java.md]: I would either switch the `FROM` to a specific version or update the Golang example above to use `latest` as well.

Comment (mlocher) [inline _pro/languages-frameworks/java.md]: Same with naming the build stages. Would be great if we could keep them the same.

**Comment by ethangj** [comment]
@rheinwein @mlocher updated the caching language, will do another update once the blog post is staged

File: _pro/builds-and-configuration/services.md [hunk 4]
```diff
@@ -181,6 +181,14 @@ app:
...
 
+### Multi-stage Builds
+
+Docker's multi-stage build feature allows you to build Docker images with multiple build stages in the Dockerfile, ultimately saving an image from just the final stage. This is great for creating "builder" workflows easily, and reducing the image size of your final Docker image.
+
+Because Codeship supports Docker natively, you will not need to do anything to get your multi-stage builds working on Codeship and we will fully support your multi-stage Dockerfiles. If you use the CLI to run builds locally, you must use CLI version 1.18 or above in order to use multi-stage bulds. Please note that using multi-stage image builds can impact the way that caching works during your Codeship Pro builds. For more information, refer to our [our caching documentation.]({{ site.baseurl }}{% link _pro/builds-and-configuration/caching.md %})
+
+You can also [read more about Docker multi-stage builds on our blog](https://blog.codeship.com/docker-17-05-on-codeship-pro/). A note that
```
Comment (mlocher) [inline _pro/builds-and-configuration/services.md]: There is a half-finished sentence in here?

Comment (rheinwein) [inline _pro/builds-and-configuration/services.md]: Which one? Did find a typo though. Will fix :)

Comment (mlocher) [inline _pro/builds-and-configuration/services.md]: This paragraph ends with _A note that_ but there is no note afterwards.

Comment (rheinwein) [inline _pro/builds-and-configuration/services.md]: Ah thanks! That just looked like GitHub cutting the sentence off in the UI.

**Pull request merged**